### ISSUE TITLE
Servo::detach() can leave the servo pin HIGH

### ISSUE THREAD

**Issue by edgar-bonet** [arduino-libraries/Servo#160]
## Platform

This issue affects the AVR platform. I have not tested other implementations of the Servo library.

## Issue description

Upon calling the `detach()` method, the Servo library stops sending pulses to the servo. This can be used as a way to instruct the servo to stop actively trying to keep a position. However, if `detach()` is called during a pulse (i.e. while the servo pin is HIGH), the pin is left in the HIGH state. This is equivalent to sending an overlong pulse, which may damage the servo.

## Expected behavior

If `detach()` is called with the pin HIGH, the current pulse should be allowed to complete, and only then should the library stop driving the pin.

## Test sketch

The sketch below demonstrates the issue. It attaches a `Servo` object to a pin, then waits for the pin to go HIGH and detaches the `Servo` object. The pin is then reliably left on the HIGH state.

```c++
#include <Servo.h>

const uint8_t servo_pin = 4;

Servo servo;

void setup() {
    Serial.begin(9600);

    // Attach the servo, then detach it while the pin is HIGH.
    servo.attach(servo_pin);
    while (digitalRead(servo_pin) == LOW) { /* wait */ }
    servo.detach();

    // Wait for the current pulse to end.
    delay(100);

    // Now make sure the pulse is really over.
    if (digitalRead(servo_pin) == HIGH)
        Serial.println("WARNING: the servo pin was left HIGH.");
}

void loop(){}
```

This sketch always prints `WARNING: the servo pin was left HIGH.`.

## Proposed solutions

A simple but racy solution would be to wait for the pin to be LOW before deactivating it:

```c++
void Servo::detach()
{
  // Do nothing if the pin is already detached. This should prevent the
  // loop below from becoming an infinite loop.
  if (!servos[this->servoIndex].Pin.isActive)
    return;

  // If the pin is currently HIGH, wait until it goes LOW before
  // deactivating it.
  while (digitalRead(servo_pin) == HIGH) { /* wait */ }
  // -- race here --

  // Now we can deactivate the pin.
  servos[this->servoIndex].Pin.isActive = false;

  // [...]
}
```

This is racy because, if the timer interrupt hits at the location marked `-- race here --`, the pin ends up deactivated in the HIGH state. A simple way to avoid the race condition would be to wait for a falling edge instead of a LOW level. Unfortunately, this could block for up to 20 ms which may be unacceptably long. A better solution might be to test the pin state and deactivate it if its LOW, all within a critical section:

```c++
void Servo::detach()
{
  do {
    // Atomically ensure the pin is LOW and deactivate it.
    ATOMIC_BLOCK(ATOMIC_RESTORESTATE) {
      if (digitalRead(servo_pin) == LOW)
        servos[this->servoIndex].Pin.isActive = false;
    }

    // If that failed, wait until the pin goes LOW.
    if (servos[this->servoIndex].Pin.isActive)
      while (digitalRead(servo_pin) == HIGH) {/* wait */}
  } while (servos[this->servoIndex].Pin.isActive);

  // [...]
}
```

Would a pull request implementing this be considered for inclusion?
